File: C++/1_Creational Patterns/2_Factory Method/Factory_Method.cpp
class Car {
public:
    virtual void accelerate() = 0;
    virtual void brake() = 0;
    virtual void steer() = 0;
};

class SportsCar : public Car {
public:
    void accelerate() override {
        // Implementation for sports car acceleration
    }

    void brake() override {
        // Implementation for sports car braking
    }

    void steer() override {
        // Implementation for sports car steering
    }
};

class Sedan : public Car {
public:
    void accelerate() override {
        // Implementation for sedan acceleration
    }

    void brake() override {
        // Implementation for sedan braking
    }

    void steer() override {
        // Implementation for sedan steering
    }
};

class SUV : public Car {
public:
    void accelerate() override {
        // Implementation for SUV acceleration
    }

    void brake() override {
        // Implementation for SUV braking
    }

    void steer() override {
        // Implementation for SUV steering
    }
};

class CarFactory {
public:
    virtual Car* createCar() = 0;
};

class SportsCarFactory : public CarFactory {
public:
    Car* createCar() override {
        return new SportsCar();
    }
};

class SedanFactory : public CarFactory {
public:
    Car* createCar() override {
        return new Sedan();
    }
};

class SUVFactory : public CarFactory {
public:
    Car* createCar() override {
        return new SUV();
    }
};

int main() {
    CarFactory* factory = new SedanFactory();
    Car* car = factory->createCar();
    car->accelerate();
    car->brake();
    car->steer();
    delete car;
    delete factory;
    return 0;
}

File: C#/1_Creational Patterns/2_Factory Method/Factory_Method.cs
public abstract class Car {
    public abstract void Accelerate();
    public abstract void Brake();
    public abstract void Steer();
}

public class SportsCar : Car {
    public override void Accelerate() {
        // Implementation for sports car acceleration
    }

    public override void Brake() {
        // Implementation for sports car braking
    }

    public override void Steer() {
        // Implementation for sports car steering
    }
}

public class Sedan : Car {
    public override void Accelerate() {
        // Implementation for sedan acceleration
    }

    public override void Brake() {
        // Implementation for sedan braking
    }

    public override void Steer() {
        // Implementation for sedan steering
    }
}

public class SUV : Car {
    public override void Accelerate() {
        // Implementation for SUV acceleration
    }

    public override void Brake() {
        // Implementation for SUV braking
    }

    public override void Steer() {
        // Implementation for SUV steering
    }
}

public interface ICarFactory {
    Car CreateCar();
}

public class SportsCarFactory : ICarFactory {
    public Car CreateCar() {
        return new SportsCar();
    }
}

public class SedanFactory : ICarFactory {
    public Car CreateCar() {
        return new Sedan();
    }
}

public class SUVFactory : ICarFactory {
    public Car CreateCar() {
        return new SUV();
    }
}

public static void Main(string[] args) {
    ICarFactory factory = new SedanFactory();
    Car car = factory.CreateCar();
    car.Accelerate();
    car.Brake();
    car.Steer();
}
